feat: upgrade react-select and make multi-select sortable

ktmud opened the pull request:
### CATEGORY

- [x] Enhancement (new features, refinement)

### SUMMARY

In order to enable [easy reordering of selected metrics/columns](https://github.com/apache/incubator-superset/issues/7129), we need to upgrade
`react-select` v1 and v3.

This PR has done a lot of cleanups and refactoring to make this happen:

1. Upgrade `react-select` from `1.3.0` to `3.1.0`.
2. Replace `react-virtualized-select` with a custom implementation based on `react-window`. Inspired by but simpler than [react-windowed-select](https://github.com/jacobworrel/react-windowed-select): we did not handle grouped options, but added typing.
3. Added **`src/components/Select`** and `SupersetStyledSelect` as the all-in-one entrypoint to all Superset styled selects. It also handles API backward compatibility for the purpose of: 1). increase usability; 2) keep changes in dependent modules minimal.
4. Virtualized list (`WindowedSelect`) will be used only when options list is long ([currently > 100](https://github.com/apache/incubator-superset/blob/3106d17edf93ad1c781131b36cf0dd5d54719ac4/superset-frontend/src/components/Select/SupersetStyledSelect.tsx#L60)).
5. Fix a couple of minor styling and UX bugs related ad-hoc metrics/filters.
6. `src/components/StyledSelect` currently used in the new ListView will be deprecated.

Since this is a fairly large change, it'd be highly appreciated if anyone who has worked on `react-select` or `SelectControl` component in Superset before could take a close look to these changes.

### BEFORE/AFTER SCREENSHOTS OR ANIMATED GIF

I know I have made a promise to keep all original styles, but the UI did change a bit in the end, especially the multi-value labels. The new default UI from `react-select` do make some sense and I don't want to add more complexity to the code just for overriding them.

Before:

![image](https://user-images.githubusercontent.com/335541/81346996-29c90c00-9070-11ea-895c-b210ef9865c3.png)

After:

![image](https://user-images.githubusercontent.com/335541/81347092-4e24e880-9070-11ea-8c3f-daf5bd8d6633.png)

Another difference is that in FilterBox, selected filter values will not have bars indicate frequency. The original design doesn't make much sense anyway since each selected value is not of equal length and it kind of interferes with legibility.

Before:

![Snip20200511_38](https://user-images.githubusercontent.com/335541/81625710-8731c580-93ae-11ea-8491-b317449ae579.png)

After:

![Snip20200511_39](https://user-images.githubusercontent.com/335541/81625718-8e58d380-93ae-11ea-94be-82cfb82d5479.png)


### TEST PLAN

CI & Manual tests.

### ADDITIONAL INFORMATION

- [x] Has associated issue: #7129 
- [x] Changes UI
- [ ] Requires DB Migration.
- [ ] Confirm DB Migration upgrade and downgrade tested.
- [x] Introduces new feature or API
- [ ] Removes existing feature or API

### REVIEWERS

@kristw @graceguo-supercat @evans @villebro @suddjian @mistercrunch @etr2460 @nytai

ktmud commented:
@nytai Feel free to continue your work on `StyledSelect`. I'll make sure to migrate the styles to `v3`. 


cc @suddjian @rusackas

nytai commented:
@ktmud thanks for taking this on. It's no easy task. Let me know if I can help!

ktmud commented on superset-frontend/cypress-base/cypress/integration/explore/control.test.js:
```diff
@@ -152,16 +148,13 @@ describe('AdhocFilters', () => {
-      cy.get('input').type('name', { force: true });
-      cy.get('.VirtualizedSelectFocusedOption').trigger('mousedown').click();
+      cy.get('.Select__control').click();
+      cy.get('input[type=text]').type('name{enter}');
```
The new `react-select` adds a `input:text` as well as `input:hidden` for form values.

kristw commented on superset-frontend/cypress-base/cypress/integration/explore/visualizations/histogram.js:
```diff
@@ -72,7 +72,6 @@ export default () =>
             comparator: 'boy',
             clause: 'WHERE',
             sqlExpression: null,
-            fromFormData: true,
```
what is this field for?

ktmud replied:
This was used to detect whether an adhoc metric/filter comes from stored metadata. It is used to automatically open the edit popover when adding a new metric/filter.

ktmud reviewed (commented):
@ All, this is finally ready for review! Apologies in advance for the giant diff.

Added some comments as pointers. For starters, the most important file to look at is https://github.com/apache/incubator-superset/pull/9628/files#diff-650ab8c9d74299035d13f427d0eaf5df

ktmud commented on superset-frontend/src/visualizations/FilterBox/FilterBox.less:
```diff
@@ -18,35 +18,10 @@
-  float: left;
-  // display: inline-block; //Ignored when float is applied. Leaving it here in case we lose the floats later.
-  padding-right: 3px;
-}
```
This could be found in `main.less` already.

ktmud commented on superset-frontend/src/visualizations/FilterBox/FilterBox.less:
```diff
@@ -18,35 +18,10 @@
-.select2-highlighted > .filter_box {
-  background-color: transparent;
-  border: 1px superset @darkest;
-}
```
`select2` is obviously not used anymore.

ktmud commented on superset-frontend/src/explore/main.less:
```diff
@@ -106,14 +98,16 @@
+        position: absolute;
+        right: 5px;
+        top: 6px;
+      }
```
This is for fixing some styles in date selector:

### Before

![image](https://user-images.githubusercontent.com/335541/81363494-5133cf00-9098-11ea-8d98-1f711b1aef55.png)


### After

![image](https://user-images.githubusercontent.com/335541/81363549-745e7e80-9098-11ea-8ca4-9eb5393b0532.png)

ktmud committed:
Fix new list view

ktmud committed:
feat: upgrade react-select v1.3.0 to v3.1.0

Upgrade `react-select`, replace `react-virtualized-select` with a custom
solution implemented with `react-window`.

Future plans include deprecate `react-virtualized` used in other places, too.

Migrate all react-select related components to `src/Components/Select`.

ktmud committed:
Fix tests

kristw commented on superset-frontend/src/components/Select/utils.ts:
```diff
@@ -0,0 +1,81 @@
+ *
+ * Always returns an array.
+ */
+export function findValue<OptionType extends OptionTypeBase>(
```
Can move flattening outside and simplify logic for always returning array.

```ts
export function findValue<OptionType extends OptionTypeBase>(
  value: ValueType<OptionType> | string,
  options: GroupedOptionsType<OptionType> | OptionsType<OptionType>,
  valueKey = 'value',
): OptionType[] {
  if (value === null || value === undefined || value === '') {
    return [];
  }

  const isGroup = Array.isArray((options[0] || {}).options);
  const flatOptions: OptionsType<OptionType> = isGroup
    ? (options as GroupedOptionsType<OptionType>)
        .flatMap(x => x.options || x);
    : (options as OptionsType<OptionType>);

  const find = (val: OptionType) => {
    const realVal = (value || {}).hasOwnProperty(valueKey)
      ? val[valueKey]
      : val;
    return (
      flatOptions.find(x => x === realVal || x[valueKey] === realVal) || val
    );
  };

  // If value is a single string, must return an Array so `cleanValue` won't be
  // empty: https://github.com/JedWatson/react-select/blob/32ad5c040bdd96cd1ca71010c2558842d684629c/packages/react-select/src/utils.js#L64
  return (Array.isArray(value) ? value : [value])
    .map(find);
}
```

ktmud replied:
Done

kristw commented on superset-frontend/src/components/Select/SupersetStyledSelect.tsx:
```diff
@@ -0,0 +1,303 @@
+    };
+
+    return (
+      // @ts-ignore (needed for passing `styles`)
```
Would declaring style as variable solve this ts issue?

kristw commented on superset-frontend/src/components/ListView/LegacyFilters.tsx:
```diff
@@ -67,9 +63,10 @@ export const FilterMenu = ({
-            onSelect={(fltr: typeof ft) =>
-              setInternalFilters([...internalFilters, fltr])
-            }
+            // @ts-ignore
```
What is the error?

ktmud replied:
> Type 'MouseEvent<{}, MouseEvent>' is missing the following properties from type '{ Header: string; id: string; value: undefined; }': Header, id, value


Inconsistent callback function signature. onSelect in MenuItem expects an event, but the callback seems to be expecting the fltr type (bad variable name btw).

Didn't want to change the actual code and didn't dig deep too deep into how to make it proper, since this file is going to be deprecated anyway (@nytai is working on a new ListView).

nytai replied:
Yes, this is old code that's going to be killed soon. I think react-bootstrap was missing proper typings when this was being implemented so some type is probably being overridden incorrectly.

nytai replied:
The issue is the typings for react-bootstrap. If EventKey is present the callback will be called with that and not a MouseEvent, but the typings don't seem to account for that. This code is all going away soon, anyway, I wouldn't spend too much time on it

ktmud commented on superset-frontend/src/components/AsyncSelect.jsx:
```diff
@@ -42,7 +43,7 @@ const defaultProps = {
 };
 
-class AsyncSelect extends React.PureComponent {
+class SupersetAsyncSelect extends React.PureComponent {
```
Probably not. Did the renaming because I was trying to `import { AsyncSelect } from src/components/Select`, as noted in the `TODO` above, but decided it was out of the scope of this PR. Can revert this change.

ktmud commented on superset-frontend/src/visualizations/FilterBox/FilterBox.jsx:
```diff
@@ -100,18 +98,15 @@ class FilterBox extends React.Component {
-  onFilterMenuOpen(chartId, column) {
-    this.props.onFilterMenuOpen(chartId, column);
+    this.onOpenDateFilterControl = (...args) => {
+      return this.onFilterMenuOpen(TIME_RANGE, ...args);
```
`this.onFilterMenuOpen` already passed `chartId` to `this.props.onFilterMenuOpen`.

ktmud committed:
Address PR comments

ktmud commented on superset-frontend/src/components/Select/utils.ts:
```diff
@@ -0,0 +1,81 @@
+  return Array.isArray(result) ? result : [result];
+}
+
+export function deepMerge(source: any, ...overrides: any) {
```
Updated.

ktmud commented on superset-frontend/src/explore/components/controls/MetricsControl.jsx:
```diff
@@ -207,51 +206,60 @@ export default class MetricsControl extends React.PureComponent {
-        nextState = { aggregateInInput: aggregate };
+    const nextState = { aggregateInInput: null };
+    const lowercaseInput = input.toLowerCase();
+    for (const aggregate of AGGREGATES_OPTIONS) {
```
Nice!

ktmud committed:
Fix a flacky Cypress test

suddjian commented on superset-frontend/src/explore/components/PropertiesModal.tsx:
```diff
@@ -236,14 +245,14 @@ function PropertiesModal({ slice, onHide, onSave }: InternalProps) {
+                isMulti
                 name="owners"
                 value={owners || []}
                 loadOptions={loadOptions}
```
Should the `loadOptions` function be returning an array of options instead of `{ options: [...] }`, like in the other `PropertiesModal`?

ktmud replied:
You are right. Thanks for catching this! I forgot to test the properties modal in the Explore view.

ktmud committed:
Adjust styles for Select in CRUD ListView

ktmud committed:
Fix loadOptions for owners select in chart PropertiesModal

TODO: add typing support for AsyncSelect props.

etr2460 commented on superset-frontend/src/explore/AdhocFilter.js:
```diff
@@ -94,7 +91,7 @@ export default class AdhocFilter {
     }
     this.isExtra = !!adhocFilter.isExtra;
-    this.fromFormData = !!adhocFilter.filterOptionName;
+    this.isNew = adhocFilter.isNew || false;
```
why not `!!adhocFilter.isNew`?

ktmud replied:
I guess both works.

etr2460 commented on superset-frontend/src/explore/constants.js:
```diff
@@ -53,10 +55,10 @@ export const HAVING_OPERATORS = [
   OPERATORS['<='],
 ];
-export const MULTI_OPERATORS = [OPERATORS.in, OPERATORS['not in']];
+export const MULTI_OPERATORS = new Set([OPERATORS.in, OPERATORS['not in']]);
```
not changed here, but I believe @villebro standardized the operators to be `IN` and `NOT IN`. Perhaps that change should be made?

ktmud replied:
Maybe in the separate PR if that's really necessary. I don't want to break things.

etr2460 reviewed (commented):
sorry for the delayed comments, forgot I had started the PR after I was ~50 files in o_O

ktmud commented on superset-frontend/cypress-base/cypress/integration/dashboard/filter.js:
```diff
@@ -53,12 +53,11 @@ export default () =>
-        .click()
-        .next()
+        .click({ force: true })
+        // parent of placeholder component
```
I think I was using a different selector when added that comment. Will delete.

ktmud commented on superset-frontend/src/components/AsyncSelect.jsx:
```diff
@@ -18,7 +18,8 @@
 import React from 'react';
 import PropTypes from 'prop-types';
-import Select from 'react-select';
+// TODO: refactor this with `import { AsyncSelect } from src/components/Select`
```
This component handles its own async logics, which doesn't seem necessary and is kind of confusing (we probably don't want to two ways of doing AsyncSelect).

ktmud commented on superset-frontend/src/components/ColumnOption.jsx:
```diff
@@ -42,28 +42,28 @@ export default function ColumnOption({ column, showType }) {
 
   return (
-    <span>
+    <div className="column-option">
```
I tend to prefer `div` unless there is a specific need to make something inline. This component doesn't seem to need inline.

ktmud commented on superset-frontend/src/components/ListView/Filters.tsx:
```diff
@@ -37,13 +44,21 @@ interface SelectFilterProps extends BaseFilter {
 const FilterContainer = styled.div`
   display: inline-flex;
-  margin-right: 8px;
+  margin-right: 1.2em;
```
Personal preference. When the spacing should scale according to font sizes, it makes more sense to use `em`s.

ktmud commented on superset-frontend/src/components/Select/SupersetStyledSelect.tsx:
```diff
@@ -0,0 +1,301 @@
+      classNamePrefix = DEFAULT_CLASS_NAME_PREFIX,
+      themeConfig,
+      stylesConfig = {},
+
```
Idk, I feel a separator line is useful, especially for those functional props. Can rearrange things to have less spaces.

ktmud commented on superset-frontend/src/components/Select/WindowedSelect/WindowedMenuList.tsx:
```diff
@@ -0,0 +1,155 @@
+ * Calculate probable option height as set in theme configs
+ */
+function detectHeight({ spacing: { baseUnit, lineHeight } }: ThemeConfig) {
+  return baseUnit * 4 + lineHeight;
```
It's the size of top and bottom paddings combined. I don't think a constant helps much here. Will add comments to make it clear.

ktmud commented on superset-frontend/src/components/Select/styles.tsx:
```diff
@@ -0,0 +1,299 @@
+export const DEFAULT_CLASS_NAME = 'Select';
+export const DEFAULT_CLASS_NAME_PREFIX = 'Select';
+
+type RecursivePartial<T> = {
```
I don't think it's used anywhere. Maybe refactor when it's needed? Unless pointed by someone, I don't think people would know it's available for use even if we have it in a global types file.

ktmud commented on superset-frontend/src/components/Select/styles.tsx:
```diff
@@ -0,0 +1,299 @@
+
+export const DEFAULT_THEME: PartialThemeConfig = {
+  borderRadius: parseInt(supersetTheme.borderRadius, 10),
+  zIndex: 11,
```
Yes, this number `11` was obtained from `variables.less` (there is a note in the type definition above). I don't know how to make a variable accessible to both less and styled components, so I just added the CSS variable name for reference. In the future, we might want to refactor all these to real CSS variables.

ktmud committed:
Address PR comments; allow isMulti in SelectControl, too

ktmud committed:
Clean up NaN in table filter values

ktmud commented:
@kristw @etr2460 This is now conflict free. Would you mind taking another look?

ktmud committed:
Fix flacky test

kristw commented:
LGTM. 
Will leave it to @etr2460.
